apis/experimental/node_classes/types.go

```diff
@@ -0,0 +1,152 @@
+package node_classes
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// This should match kubeadm's API as closely as possible, so its definition is
+// uninteresting and not fleshed out.
+//
+// What is more important (for this PR) is the representation of nodes and how
+// they are linked to the concept of the Cluster.
+type Cluster struct {
```

Comment:
Do we want the top level cluster object to include both the node definitions and the control plane definition to make it easier for people to read in a yaml file? That would mean that we would add a ControlPlane struct and embed it along with an [] of node pools in the Cluster struct.

Reply:
With your recent suggestion of scoping the Cluster API to be a single cluster, and letting the Cluster Registry be the de facto mechanism for tracking multiple clusters, I'm not sure it would offer mechanical value (since all node pools found on a cluster would correspond to that cluster). Do you think a multidoc yaml file is reasonable enough to get the full picture of a cluster without adding literal pointers in the structs?

What makes me hesitate here is that if someone wants to keep N different clusters reasonably in sync across different cloud providers or on-premise installations (using the same versions of Kubernetes, the same features, the same control plane flags, etc.), it would be nice if there were a limited set of API objects they try to keep completely in sync. This may not end up being feasible, but the fewer "ignore this particular field, which is very likely to differ from one cluster to the next" cases we have, the better I think. Linking the Cluster to the NodePools explicitly immediately gives us this case to ignore.

Reply:
>Do you think a multidoc yaml file is reasonable enough to get the full picture of a cluster without adding literal pointers in the structs?

Does that mean if we have a yaml file with N cluster, there can be <= N control plane struct definitions in the file. 

What I understand is a sample yaml file would look something like this : 

```
kind: Cluster
metadata:
    name: clusterA
spec:
    nodePool:
    . . . .

----- 
kind: ControlPlane
metadata:
    clusterName: ["clusterA"]
spec:
    . . . .

----- 
kind: Cluster
metadata:
    name: clusterB
spec:
    nodePool:
    . . . .

----- 
kind: Cluster
metadata:
    name: clusterC
spec:
    nodePool:
    . . . .

----- 
kind: ControlPlane
metadata:
    clusterName: ["clusterB", "clusterC"]
spec:
    . . . .

```